modules/deriving/src/test/scala/shapeless3/deriving/deriving.scala

```diff
@@ -393,3 +393,25 @@ class DerivationTests:
     assertEquals(Right(ISB(42, "foo", true)), parser.parseShort("s=foo,i=42,b=true,hidden=?"))
     assertEquals(Left("Missing field 's';"), parser.parseShort("i=42,b=kinda"))
     assertEquals(Left("Invalid field 'broken';"), parser.parseShort("i=42,broken,?"))
+
+  // Testing compilation only
+  def userDefinedMirrors(): Unit =
+    def show[A](using
+        Mirror.ProductOf[A] {
+          type MirroredLabel = "ISB"
+          type MirroredElemLabels = ("i", "s", "b")
+          type MirroredElemTypes = (Int, String, Boolean)
+        }
+    ) = Show[A]
+
+    def bifunctor[F[_, _]](using
+        K2.Refl[Mirror.Sum, F] {
+          type MirroredElemTypes[A, B] = (Either[A, B], (A, B))
+        }
+    ) = Bifunctor[F]
+
+    def functorK[Alg[_[_]]](using
+        K11.Refl[Mirror.Product, Alg] {
+          type MirroredElemTypes[F[_]] = (F[Int], String)
+        }
+    ) = FunctorK[Alg]
```

Comment:
I don't see how this is actually testing the issue from #218. By refining `Refl`, your parameter is still asking whoever is calling the function to provide a `Mirror { type Kind = ... }`, i.e. a shapeless-specific `ProductGeneric`, exactly the thing that we want to avoid. I think a more realistic test would be something like this:

```scala
class Foo(val i: Int, val s: String, val b: Boolean) extends Product {
   ... Product impl here
}

inline given Mirror.ProductOf[Foo] {
   // I expect that this will not work without spelling out all the overriden types in the given's type declaration and making it inline
   type MirroredType = Foo
   type MirroredMonoType = Foo
   type MirroredLabel = "Foo"
   type MirroredElemTypes = (Int, String, Boolean)
   type MirroredElemLabels = ("i", "s", "b")
} = new Mirror {
   // Note: No mention of `type Kind`, ProductGeneric or any other Shapeless specifics
   type MirroredType = Foo
   type MirroredMonoType = Foo
   type MirroredLabel = "Foo"
   type MirroredElemTypes = (Int, String, Boolean)
   type MirroredElemLabels = ("i", "s", "b")
   override def fromProduct = ...
}

summon[Show[Foo]] // should internally call K0.mkProductInstances, thus triggering an implicit search for K0.ProductGeneric[Foo]
```
and accordingly for `K2`, `K11`, and so on.

Reply:
I guess I can change the test to use `Mirror` directly, it will be a bit more verbose